src/controllers/api-key.js

```diff
@@ -10,82 +10,259 @@
+      // active API keys for the given imsOrgId.
+      const apiKeys = await dataAccess.getApiKeysByImsUserIdAndImsOrgId(imsUserId, imsOrgId);
+
+      const validApiKeys = apiKeys.filter(
+        (apiKey) => apiKey.isValid(),
+      );
+
+      // Check if the user has reached the maximum number of API keys.
+      // Currently, we only allow 3 API keys per user.
+      if (validApiKeys && validApiKeys.length >= maxApiKeys) {
+        throw new ErrorWithStatusCode(`Invalid request: Exceeds the limit of ${maxApiKeys} allowed API keys`, STATUS_FORBIDDEN);
+      }
+
+      const { email } = imsUserProfile;
+      let username;
+      if (hasText(email)) {
+        [username] = email.split('@');
+      }
 
-    validateRequestData(data);
-    validateImsOrgId(imsOrgId);
+      // Create the API key
+      const apiKey = username ? `${username}-${crypto.randomUUID()}` : crypto.randomUUID();
+      const hashedApiKey = hashWithSHA256(apiKey);
 
-    if (data) {
-      const error = new ErrorWithStatusCode('Create API key not implemented', 501);
+      let scopes = {};
+
+      // In response to an 'imports' feature request, we set the scopes to
+      // imports.read and imports.write
+      if (data.features.includes('imports')) {
+        scopes = [
+          {
+            name: 'imports.read',
+          },
+          {
+            name: 'imports.write',
+            domains: data.domains,
+          },
+        ];
+      }
+
+      const expiresAt = new Date();
+      expiresAt.setMonth(expiresAt.getMonth() + 6);
+
+      const apiKeyEntity = await dataAccess.createNewApiKey({
+        id: crypto.randomUUID(),
+        name: data.name,
+        scopes,
+        createdAt: new Date().toISOString(),
+        expiresAt: expiresAt.toISOString(),
+        imsOrgId,
+        imsUserId,
+        hashedApiKey,
+      });
+
+      return createResponse(ApiKeyDto.toJSON(apiKeyEntity, apiKey), STATUS_CREATED);
+    } catch (error) {
+      log.error(`Failed to create a new api key: ${error.message}`);
       return createErrorResponse(error);
     }
-    return createResponse({}, 501);
   }
 
   /**
    * Delete an API key.
    * @param {Object} requestContext - Context of the request.
-   * @returns {Promise<Response>} 204 No Content.
+   * @returns {Promise<Response>} - 204 No Content.
    */
   async function deleteApiKey(requestContext) {
     const { pathInfo: { headers }, params: { id } } = requestContext;
-    const imsOrgId = headers['x-ims-gw-org-id'];
+    const imsOrgId = headers['x-gw-ims-org-id'];
+
+    try {
+      const imsUserToken = getImsUserToken(headers);
+      await validateImsOrgId(imsOrgId, imsUserToken);
+      const apiKeyEntity = await dataAccess.getApiKeyById(id);
+      const { authInfo: { profile } } = attributes;
 
-    validateImsOrgId(imsOrgId);
-    if (id) {
-      const error = new ErrorWithStatusCode('Delete API key not implemented', 501);
+      // Currently the email is assigned as the imsUserId
```

Comment:
Let's match this comment to the other instance.

Reply:
Another idea: create a `getImsUserIdFromProfile(profile)` function and call it from both places, that way they can't so easily get out of sync.